keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -162,10 +163,63 @@ TODO: complete this list
 
 ### Scale Targets for GA
 
-* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+The scale target for GA of custom resources are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details).
+
+The targets are defined by the below suggested maximum limits, which are organized the same way as the [Kubernetes native type thresholds](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds), with only one change:
+
+- Since custom resources can be arbitrarily large, we have broken down the limit by custom resource object size.
+
+**Custom Resources per Definition:**
+
+| Object size | Suggested Maximum Limit: scope=namespace | Suggested Maximum Limit: scope=cluster |
+| --- | --- | --- |
+| 10kb | 2500 | 12500 |
+| 25kb | 1000 | 5000 |
+| 50kb | 500 | 2500 |
+
+Since, in practice, custom resources scale farther without conversion webhooks
+within the SLI/SLOs (roughly 2x according to our scale tests), custom resource
+definition authors should be careful to adhere to these limits, so that a in the
+future, a webhook converter may safely be added as part of a custom resource
+version upgrade.
+
+_Note: For custom resources of custom resource definitions using `scope: Namespaced`: the scope=namespace
+suggested maximum limit indicates how many custom resource objects may be in each namespace,
+and the scope=cluster suggested maximum limit indicates how many custom resource objects may
+be in the cluster total. For custom resources of custom resource definitions using `scope: Cluster`: only
+the scope=cluster suggested maximum limit applies._
+
+**Custom Resource Definitions:**
+
+| Suggested Maximum Limit: scope=cluster |
+| --- |
+| 500 |
+
+_Note: The Custom Resource Definition suggested maximum limit was selected not
+due to the above SLI/SLOs, but instead due to the latency OpenAPI publishing,
+which is a background process that occurs asychroniously each time a Custom
+Resource Definition schema is updated. For 500 Custom Resource Definitions it takes
+slightly over 35 seconds for a definition change to be visible via the OpenAPI
+spec endpoint._
+
+Given that the performance and scalability of conversion webhooks are the
+responsibility of their author, Custom resource scale targets are apply only for
+conversion webhooks that are within the follow latencies for the above suggested
+maximum limits:
+
+| Conversion Webhook 99th percentile latency: scope=namespace | Conversion Webhook 99th percentile latency: scope=cluster |
+| 2.5 seconds | 15 seconds |
```

Comment:
I don't get it - how we can meet an SLO of 1s with webhook taking 15 seconds?

Reply:
15 seconds is for scope=cluster, which has a 30 second SLO (https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md). I'll organize the table a bit different to make this more clear.

Reply:
But it's only for lists - which requires converting N objects.
Unless I'm completely missing something, you need to call a convertion webhook for every single returned objects. And for a single object, the SLO is 1s.

Reply:
Ah, conversion webhooks accept a list of objects and return a list of objects.

Reply:
So let me understand it.
Let's say, I'm calling: "GET my-small-CR-object". IIUC, this still requires calling coversion webhook, right? If so, the SLO for it is 1s, so we need to have an additional guarantee for it, at least.

[BTW, using half of the slo threshold seem high to me.]

Reply:
> So let me understand it.
> Let's say, I'm calling: "GET my-small-CR-object". IIUC, this still requires calling coversion webhook, right? If so, the SLO for it is 1s, so we need to have an additional guarantee for it, at least.

Yes, I've added reorganized the table with a scope=resource for single CR SLO:

```
 | scope | Expected conversion Webhook SLO: 99th percentile latency |                                                                                                                                                                                          
 | --- | --- |                                                                                                                                                                                                                                                 
 | scope=resource | 500ms |                                                                                                                                                                                                                                    
 | scope=namespace | 2.5 seconds |                                                                                                                                                                                                                             
 | scope=cluster | 15 seconds |    
```

> [BTW, using half of the slo threshold seem high to me.]

This is based on our benchmarks showing that we scale to almost exactly 2x our thresholds without CRD conversion as we do with CRD conversion. @roycaihw is still doing a couple additional benchmarks on it, so if we get updated data we'll apply it here.

Reply:
I'm against those limits.

Because what that means is that SLO on the work that has to happen on our side is then basically 0.5s (for individual resource).
Given that CRDs on our side aren't cheaper (in fact they are even more expensive because they don't support protobufs, etc.), that's not something we will satisfy.

Reply:
I assume the expectation of "webhook SLO being half of CR SLO" comes from 
> (roughly 2x according to our scale tests)

however, the increase in latency consists of 
1. apiserver serializes the CRs into a `ConversionReview`
2. apiserver sends the `ConversionReview` to the webhook and the webhook de-serializes the request
3. the webhook performs the conversion (which was no-op in our scale tests)
4. the webhook serializes the `ConversionReview` and sends it back to apiserver
5. apiserver deserializes the `ConversionReview`

IIUC No.1 and No.5 should not be considered part of webhook SLO according to [[1]](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details), so I would expect webhook SLO to be shorter

Reply:
> IIUC No.1 and No.5 should not be considered part of webhook SLO according to [1]

That being said, in our scale tests we included the client deserialization time in the LIST p99 latency, which also doesn't belong to apiserver SLO by definition. Therefore apiserver actually has some more room meeting the SLO

Reply:
@roycaihw Oh, you're right. If step #1 and step #4 have the same cost, and step #2 and step #5 have the same cost, then I'd expect the wehbook SLO to 1/2 of the latency of all the steps, or 1/4 the total SLO. Does that math sound right? Should be confirm experimentally?

Reply:
> That being said, in our scale tests we included the client deserialization time in the LIST p99 latency, which also doesn't belong to apiserver SLO by definition. Therefore apiserver actually has some more room meeting the SLO

@wojtek-t You had said something about this before, right? Do we currently include client serialization/deserialization in the native type SLOs?

Reply:
> that SLO on the work that has to happen on our side is then basically 0.5s (for individual resource)

Reading/writing individual resource is fast according to the scale test results. I'm confident that we can meet the apiserver SLO even if we give 0.5s to the webhook.

Reply:
> I'd expect the wehbook SLO to 1/2 of the latency of all the steps, or 1/4 the total SLO. Does that math sound right? Should be confirm experimentally?

That's my guess. I'm a little concerned about the network latency in No.2 and No.4. We have pprof analysis but it doesn't show the IO latency in webhook conversion. I will add some instrumentation today

Reply:
LIST request sample (object size: 10kb; object count: 2500) from pprof trace:

Goroutine Name: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1

Goroutine | Total | Execution| Sync block
-- | -- | -- | --
300285053 | 3373ms | 1989ms | 1337ms
300285335 | 3298ms | 1909ms | 1339ms
300285765 | 3252ms | 1820ms | 1375ms

The 1337ms sync block corresponds to [k8s.io/client-go/rest.(*Request).Do](https://github.com/kubernetes/kubernetes/blob/69a34f6a6f67de47cb9b72b6ac98e089d301beb3/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L184), which is pretty much "from sending the first byte to the webhook to receiving the last byte from the webhook"

![webhook-latency-trace](https://user-images.githubusercontent.com/6629765/61911921-aaf25080-aeed-11e9-9bb0-c05771c536bc.png)
1/4 of the total 5s SLO -> 1.25s is slightly shorter than a noop webhook convertor's latency (1.3-1.4s) in this case. I'm going to sample other size/count settings and add the results to the google doc

Reply:
So let me clarify my disagreement from the above (because I don't think any of the above comments really change any of my thinking).

> @wojtek-t You had said something about this before, right? Do we currently include client serialization/deserialization in the native type SLOs?

correct - we start measuring from when the request reaches apiserver (well, roughly, because handshaking etc. is not included there) And we stop when the "write" operation of writing result to user is done (which basically includes client tcp-acking receiving the response).
The reason is that this is basically closer to what users really see and we want the SLOs to be user-oriented.


Going back to my original agreement - I don't think any benchmark data is going to change anything. The experiments you're doing are artificial benchmarks where you're not running any other load against the apiserver and etcd. And that is pretty crucial.

So let me be more specific - in our 5k tests, there are resources for which 99th percentile of latency is exceeding 0.5s. So if for those we would add 0.5s of webhook, we would exceed 1s SLO and we are immediately out of SLO.
And there is no reason why this standard work we're doing for CRs (like writing to etcd, additional processing in apiserver, etc.) would be shorter for CRs - in fact it will be even longer due to: json vs protobuf etc.

So to summarize what I'm saying:
- if you allow webhook to take 0.5s at 99th percentile, there is NO WAY we will be able to satisfy 1s SLO in large and busy enough clusters.

Reply:
So either we will still be saying in SLO that all time related to webhooks is excluded from the latency, or if we would like to include that (which in case of conversion webhooks makes some sense, because they are kind-of part of basic functionality), the value has to be much much smaller.

Reply:
@wojtek-t You're right. Giving the webhook 0.5s of the 1s total for the 99th percentile doesn't work. And I don't think we have numbers we're totally happy with, right now we're thinking something in this range:

 | scope | Expected conversion Webhook SLO: 99th percentile latency |                                                                                                                                                                                          
 | --- | --- |                                                                                                                                                                                                                                                 
 | scope=resource | 250ms |                                                                                                                                                                                                                                    
 | scope=namespace | 1.25 seconds |                                                                                                                                                                                                                            
 | scope=cluster | 7.5 seconds |

But for scope=resource this might still be too high?  For the namespace and cluster scopes we have the flexibility of adjusting the threshold to keep things in SLO, but for scope=resource that's not an option..

It sounds like we're in agreement that we need to include conversion webhook latency in CR SLOs, which intuitively makes from the users perspective-- they might need to introduce conversion at any time and introducing conversion shouldn't drop them out of SLO. Our targets here now are centered around SLOs for CRs with conversion webhooks plus a SLO the webhooks should adhere too.

For the webhook SLOs we're focusing on what the webhook author "sees", so we're not counting the serialization/deserialization cost on the apiserver to send/recieve the webhook request.

Reply:
> It sounds like we're in agreement that we need to include conversion webhook latency in CR SLOs, which intuitively makes from the users perspective-- they might need to introduce conversion at any time and introducing conversion shouldn't drop them out of SLO.

Agree.

> For the webhook SLOs we're focusing on what the webhook author "sees", so we're not counting the serialization/deserialization cost on the apiserver to send/recieve the webhook request.

This is actually a good point - thanks for making that explicit (in my mental model I was kind-of treating it as e2e from apiserver, which I agree is a wrong model.

> But for scope=resource this might still be too high? 

Definitely. Especially given the above, which translates (based on the description from couple comments above) to only points 3, right?

So if you agree with the above, I personally think that the SLO should be fairly small. I was initially thinking about 5% for webhook, but that is probably a bit too aggressive.

So let's say that we're able to spend 20% of the SLO on webhook-related stuff. However, that 20% should contain all operations related to webhook (including serialization and sending it sending necessary info). I think this is the value I can live with.

That said, it still translates for something much faster from user POV - I would say that we should live something like 5% for real user-logic (i.e. from the moment the object is deserialized, to the moment it's ready for serializing).
While 50ms isn't a lot, it's order(s) of magnitude more we're using for conversion for internal types. So I think it's actually fine.

With all the above I realized, that imho we should also include the condition that RTT between kube-apiserver and webhook provider should be fairly small.

Reply:
Thanks @wojtek-t , I'll update the PR to match.

Reply:
Okay, @roycaihw helped me go through the numbers and we've updated this PR with:

CRs per CRD:

| Object size | Suggested Maximum Limit: scope=namespace (5s p99 SLO) | Suggested Maximum Limit: scope=cluster (30s p99 SLO) |                                                                                                                                
 | --- | --- | --- |                                                                                                                                                                                                                                           
 | 10kb | 1250 | 7500 |                                                                                                                                                                                                                                        
 | 25kb | 500 | 3000 |                                                                                                                                                                                                                                         
 | 50kb | 250 | 1500 |

Conversion webhooks:

| scope | Expected conversion Webhook SLO: p99 latency |                                                                                                                                                                                                      
 | --- | --- |                                                                                                                                                                                                                                                 
 | resource | 50ms |                                                                                                                                                                                                                                           
 | namespace | 1 seconds |                                                                                                                                                                                                                                     
 | cluster | 6 seconds | 

These numbers give much more headroom under heavy load (e.g. 5k scale tests) and put the ratio of the webhook latency to total request latency to 20%.

Reply:
how much headroom do we have under the 10kb/7500 cluster limit? ideally we could accomodate 1 CR per namespace, which https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md specifies as 10000 cluster-wide

Reply:
> how much headroom do we have under the 10kb/7500 cluster limit?

Oh that's a good point. Dropping these limits put us below the 1 CR per namespace target. Our benchmarks found a limit of 15000 for 10kb (with conversion webhook), We'd reduced them to both give head room to account for the heavy load that can occur during 5k scale tests and to ensure we offer conversion webhook authors a SLO they can viably meet. We basically cut our limits in half to do this. But we do have wiggle room now, and our survey results explicitly called out 1 CR per namespace as being the count for a variety of CRs, so I'm in favor for pushing up our targets for this.  After all, these numbers are suppose to be goals, not just a summary of current system limits.

Reply:
Here's the bump I've made. I've adjusted all our numbers (rounded to decimal numbers with lots of zeros) to stay consistent.

| Object size | Suggested Maximum Limit: scope=namespace (5s p99 SLO) | Suggested Maximum Limit: scope=cluster (30s p99 SLO) |                                                                                      
| --- | --- | --- |                                                                                                                                                                                                 
| 10kb | 1500 | 10000 |                                                                                                                                                                                             
| 25kb | 600 | 4000 |                                                                                                                                                                                               
| 50kb | 300 | 2000 |